manifests/server/install.pp

```diff
@@ -33,7 +33,7 @@
                                       default => ['puppetmaster-common', 'puppetmaster'],
                                     },
         /^(FreeBSD|DragonFly)$/ => [],
-        default                 => ['puppet-server'],
+        default                 => ['puppetserver'],
       },
       'puppetserver' => 'puppetserver',
```

Comment:
shouldn't this be used automatically as `$::puppet::server::implementation` should be set accordingly?

Reply:
I believe this IS where $::puppet::server::implementation is defined:

$server_package_default = $::puppet::server::implementation ? {
      'master'       => $::osfamily ? {
        'Debian'                => $puppet4 ? {
                                      true    => ['puppet-master'],
                                      default => ['puppetmaster-common', 'puppetmaster'],
                                    },
        /^(FreeBSD|DragonFly)$/ => [],
        default                 => ['puppet-server'],
      },
      'puppetserver' => 'puppetserver',

Reply:
nope.
`$server_package_default` is deduced from `$::puppet::server::implementation`, which in turn should be `puppetserver`, if Puppet 4 AIO packages are used.

```puppet
    $server_package_default = $::puppet::server::implementation ? {
      'master'       => $::osfamily ? {
[...]
      'puppetserver' => 'puppetserver',
    }
```

Reply:
Right. IF puppet 4 AIO packages are used BEFORE installing the foreman-installer.rpm (which is not noted in the quick start guide.) If puppet 3 packages are used, the installer fails because the forman-installer.rpm requires puppet v4's puppet-agent.rpm package, but cascades no dependency for puppetserver.rpm (v4), so this package doesn't install if you do a `yum -y install katello` . This means that puppet-server.rpm installation isn't attempted until after you run `foreman-installer --scenario katello` at which point this puppet module defaults to and tries to install puppet-server.rpm (v3) causing a dependency conflict with puppet-agent.rpm. if you have Puppet 3 installed, the foreman-installer.rpm fails to install because it can't find the puppet-agent.rpm dependency. Therefore the only way (and proper way) to fix this is to remove all support for v3, which isn't going to work anyway.

Go ahead and give a fresh installation (following the quick start guide exactly) a try if you aren't sure what I'm blathering on about - you can't miss it.

Reply:
Sounds like an RPM issue outside the scope of the puppet-puppet module then?

Reply:
looking at the redmine issue: 1.16 and 1.17 indeed do not work with Puppet 3.x, while 1.15 should.

Reply:
So, it sounds like this needs to be merged for 1.16 and 1.17, and another bug filed against 1.15 to have the RPM dependencies fixed. If it's just working with a spec file, I can probably commit that update, but if the spec files for packaging are generated out of ruby, I'm sure to get lost pretty quickly...

Reply:
see https://github.com/theforeman/foreman-packaging, the RPMs are generated from their respectibe branches i.e. rpm/1.15, rpm/1.16, rpm/develop (the later for 1.17/nightly)

Reply:
Regarding "this needs to be merged for 1.16 and 1.17": I still think the current puppet-puppet code is fine as-is by the explanation above.

Reply:
Sorry, I am having a hard time following this. Here is what I know.

For 1.15, puppet 3 and puppet 4 are supported. The foreman-installer package, in 1.15., properly handles that (https://github.com/theforeman/foreman-packaging/blob/rpm/1.15/foreman-installer/foreman-installer.spec#L19).

For 1.16 and 1.17, only Puppet 4 is supported. 

I think perhaps this puppet module (which is treated as providing stand alone puppet management support and is also used by the installer) is still striving to manage both Puppet 3 and Puppet 4/5.

The quickstart guide is out of date and should be updated. Please file a docs bug or a PR to help us fix it up! Note that the quickstart guide does not apply to Katello which I noticed you mentioned in there. In the case of Katello, this guide is the only guide we support (https://theforeman.org/plugins/katello/nightly/installation/index.html). And we are aware that this varies from Foreman installs and are working to bring these two in line.

Does this help @jdshewey ? I'd like to work to ensure there are as few holes for users to fall into as possible with installs.

Reply:
Some. I see this requires puppet 3:

    Requires:   curl
    Requires:   %{?scl_prefix}puppet >= 3.6.0

But I'm not seeing how this allows for the installation of puppet-agent, the new name of puppet in version 4. Secondly, I do not see where this installer requires the "puppet-server" rpm (v3) or "puppetserver" (v4) rpm, which are also required dependencies. From the perspective of the RPM, this is the problem, unless there is some code I missed. This code is used by all scenarios for the installer, correct?

As for Katello, am I correct in inferring from your link that only the nightly unstable build is supported? Or can Katello be installed from 1.15?

Reply:
Also, Doesn't the nightly systest use vagrant or ansible without actually walking through installing the RPMs and walking through forman-install --scenario katello?

Reply:
Also, @mmoll if you think it is fine as is, why don't you try giving the install a go for yourself. It most decidedly runs into issues.

Reply:
This appears to have been reported by someone else [in January as well](https://github.com/theforeman/puppet-puppet/issues/472)

Reply:
The referenced issue has nothing to do with the choice which puppet master/server package is chosen for install, only about the RPM/DEB version of the respective package and the puppetserver configfile format.